README.md
Introduction
------------

[![changelog](http://allmychanges.com/p/python/django-fields/badge/)](http://allmychanges.com/p/python/django-fields/?utm_source=badge)

Django-fields is an application which includes different kinds of models fields.

Right now, application contains two fields with encryption support:
EncryptedCharField and EncryptedTextField.


How to run tests
----------------

Examples can be found at the `examples` directory. Look at the, `tests.py`.
Same project is used to run unittests. To run them, just fire `../env/bin/python ./run-tests.py`.

Contributors
------------

* [zbyte64](http://www.djangosnippets.org/users/zbyte64/) — thanks to for
  his [django snippet](http://www.djangosnippets.org/snippets/1095/) for encrypted
  fields. After some fixes, this snippet works as supposed.
* John Morrissey — for fixing bug in PickleField.
* Joe Jasinski — different fixes and new fields for encripted email and US Phone.
* Colin MacDonald — for many encripted fields added.
* Igor Davydenko — PickleField.
* kromem - Added support for specifying `block_type` on encrypted fields.
* Brooks Travis - new field for encrypted US Social Security Number and other fixes.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.5.4](https://github.com/last-partizan/django-fields/compare/v0.5.4-beta.2...v0.5.4) (2022-05-27)

### [0.5.4-beta.2](https://github.com/last-partizan/django-fields/compare/v0.5.4-beta.1...v0.5.4-beta.2) (2022-05-27)

### [0.5.4-beta.1](https://github.com/last-partizan/django-fields/compare/v0.5.4-beta.0...v0.5.4-beta.1) (2022-05-27)

### [0.5.4-beta.0](https://github.com/last-partizan/django-fields/compare/v0.5.3...v0.5.4-beta.0) (2022-05-27)


### Bug Fixes

* Fix release syntax ([e931c53](https://github.com/last-partizan/django-fields/commit/e931c53ba8832308d18d851bba729245fa3a1208))

### [0.5.3](https://github.com/last-partizan/django-fields/compare/v0.5.2...v0.5.3) (2022-05-27)

### [0.5.2](https://github.com/last-partizan/django-fields/compare/v0.5.1...v0.5.2) (2022-05-14)

### [0.5.1](https://github.com/last-partizan/django-fields/compare/v0.5.0...v0.5.1) (2022-05-14)


### Features

* Use pytest ([01f1448](https://github.com/last-partizan/django-fields/commit/01f1448cc932145cd2f97fcbd0ef56e2a0a4367f))

## [0.5.0](https://github.com/last-partizan/django-fields/compare/v0.4.0...v0.5.0) (2022-05-14)


### Features

* Use github-ci ([70f2172](https://github.com/last-partizan/django-fields/commit/70f21721cf60cecc8c3fd0ed042bd5aab882ae09))


### Bug Fixes

* Remove six dependency ([5ca4186](https://github.com/last-partizan/django-fields/commit/5ca41861e31ddb5ff68ddb2c6c4112a2e6f55fab))

ChangeLog
=========

0.3.0 (2014-09-12)
------------------
	
* Added: Support for Django migrations (>=1.7) thanks to `Field.deconstruct()`.

0.2.3
-----

* Added: Ability to specify `secret_key` as an argument to the field constructor (via [markkennedy](https://github.com/svetlyak40wt/django-fields/pull/40 "Issue #40")).
* Added: `EncryptedUSPhoneNumberField` and `EncryptedUSSocialSecurityNumberField` now try to import from the standalone django-localflavor before falling back to the version included in Django (this is necessary to support Django 1.6 and newer) (via [mjacksonw](https://github.com/svetlyak40wt/django-fields/pull/36 "Issue #33")).

0.2.2
-----

* Fixed: Django admin needs to be able to create blank instances of the fields in order to create a new model. This broke with `BaseEncryptedNumberField`. (via [defrex](https://github.com/svetlyak40wt/django-fields/pull/32 "Issue #32"))
* Fixed: `block_type` wasn't added to the south rules. (via [defrex](https://github.com/svetlyak40wt/django-fields/pull/33 "Issue #33"))
* Fixed: Newer code paths with `block_type` specified couldn't reuse the `cipher` object on the field class. `to_python` was already redefining it before decrypting the value, but `get_db_prep_value` wasn't before encrypting. The first time you used a model it would be fine, but the second would fail. Thus the tests were passing but the classes were functionally useless in an application. (via [defrex](https://github.com/svetlyak40wt/django-fields/pull/34 "Issue #34"))

0.2.1
-----

* Added: `EncryptedUSSocialSecurityNumberField`, which handles the special-case logic of validating and encrypting US Social Security Numbers, using `django.contrib.localflavor.us.forms.USSocialSecurityNumberField`. (via [Brooks Travis](https://github.com/svetlyak40wt/django-fields/pull/24 "Pull Request 24"))
* Fixed: Issue [#21](https://github.com/svetlyak40wt/django-fields/issues/21 "Issue #21").
* Changed: `django_fields.fields.BaseEncryptedField` now supports specification of cipher `block_type` via keyword argument. (via [kromem](https://github.com/svetlyak40wt/django-fields/pull/26 "Pull Request 26"))
* Added: Deprecation warning for fields that do not specify a `block_type`.

0.2.0
-----

* Added: Class `django_fields.models.ModelWithPrivateFields`, which allows to use private fields, starting from two underscores.
* Fixed: `BaseEncryptedDateField.get_db_prep_value` errors.
* Changed: Now virtualenv is used for test enviroment. Buildout.cfg was removed.

0.1.3
-----

* Fixed: `EOFError` handling in `PickleField`.
* Changed: Buildout file was changed to test against Django 1.2.5 and include `PyCrypto`.

0.1.2
-----

* Added: `EncryptedEmail` and `USPhoneNumber` fields.
* Added: `EncryptedNumber` fields.
* Added: `EncryptedDateTimeField`.
* Added: `EncryptedDateField` class.
* Added: South support.
* Added: Unit tests and associated utility functions.
* Fixed: Deprecation warnings related to the settings in example project.
* Fixed: Deprecation warnings, related to `get_db_prep_vasue`.
* Fixed: Edge case in encryption consistency.
* Changed: `EncryptedCharField` now enforces max length.

0.1.1
-----

* Added: `PickleField` class.
* Added: Encrypt field.
* Added: Buildout config and example application.
* Added: `setup.py` and `MANIFEST.in.`
* Fixed: Issue #1 - "`EncryptedCharField` raises a traceback in the django admin".
* Fixed: `max_length` issue.
* Changed: Now `__import__` compatible with python 2.4 in `BaseEncryptedField.__init__`.
* Changed: Code was moved to `src`.
* Changed: Get rid of custom string class. It was replaced with string prefix.
* Changed: Settings were changed to test with mysql.


src/django_fields/fields.py
import binascii
import codecs
import datetime
import string
import sys
import warnings
import pickle

from django import forms
from django.core.validators import EMPTY_VALUES
from django.db import models
from django.conf import settings
from django.utils.translation import gettext_lazy as _
from django.utils.encoding import force_str
from Cryptodome import Random
from Cryptodome.Random import random

DEFAULT_BLOCK_TYPE = 'MODE_ECB'


class BaseEncryptedField(models.Field):
    '''This code is based on the djangosnippet #1095
       You can find the original at http://www.djangosnippets.org/snippets/1095/'''

    def __init__(self, *args, **kwargs):
        self.cipher_type = kwargs.pop('cipher', 'AES')
        self.block_type = kwargs.pop('block_type', None)
        self.secret_key = kwargs.pop('secret_key', settings.SECRET_KEY)
        self.secret_key = self.secret_key[:32]
        self.secret_key = bytes(self.secret_key.encode('utf-8'))

        if self.block_type is None:
            warnings.warn(
                "Default usage of pycrypto's AES block type defaults has been "
                "deprecated and will be removed in 0.3.0 (default will become "
                "MODE_CBC). Please specify a secure block_type, such as CBC.",
                DeprecationWarning,
            )
            self.block_type = DEFAULT_BLOCK_TYPE
        try:
            imp = __import__('Cryptodome.Cipher', globals(), locals(), [self.cipher_type], -1)
        except:
            imp = __import__('Cryptodome.Cipher', globals(), locals(), [self.cipher_type])
        self.cipher_object = getattr(imp, self.cipher_type)
        if self.block_type != DEFAULT_BLOCK_TYPE:
            self.prefix = '$%s$%s$' % (self.cipher_type, self.block_type)
            self.iv = Random.new().read(self.cipher_object.block_size)
            self.cipher = self.cipher_object.new(
                self.secret_key,
                getattr(self.cipher_object, self.block_type),
                self.iv)
        else:
            self.cipher = self.cipher_object.new(self.secret_key, getattr(self.cipher_object, self.block_type))
            self.prefix = '$%s$' % self.cipher_type

        self.original_max_length = max_length = kwargs.get('max_length', 40)
        self.unencrypted_length = max_length
        # always add at least 2 to the max_length:
        #     one for the null byte, one for padding
        max_length += 2
        mod = max_length % self.cipher_object.block_size
        if mod > 0:
            max_length += self.cipher_object.block_size - mod
        if hasattr(self, 'iv'):
            max_length += len(self.iv)
        kwargs['max_length'] = max_length * 2 + len(self.prefix)

        super(BaseEncryptedField, self).__init__(*args, **kwargs)

    def _is_encrypted(self, value):
        return (
            isinstance(value, str) and
            value.startswith(self.prefix)
        )

    def _get_padding(self, value):
        # We always want at least 2 chars of padding (including zero byte),
        # so we could have up to block_size + 1 chars.
        mod = (len(value) + 2) % self.cipher_object.block_size
        return self.cipher_object.block_size - mod + 2

    def from_db_value(self, value, expression, connection):
        if self._is_encrypted(value):
            if self.block_type != DEFAULT_BLOCK_TYPE:
                self.iv = binascii.a2b_hex(value[len(self.prefix):])[:len(self.iv)]
                self.cipher = self.cipher_object.new(
                    self.secret_key,
                    getattr(self.cipher_object, self.block_type),
                    self.iv)
                decrypt_value = binascii.a2b_hex(value[len(self.prefix):])[len(self.iv):]
            else:
                decrypt_value = binascii.a2b_hex(value[len(self.prefix):])
            return force_str(
                self.cipher.decrypt(decrypt_value).split(b'\0')[0]
            )
        return value

    def get_db_prep_value(self, value, connection=None, prepared=False):
        if value is None:
            return None

        value = bytes(value.encode('utf-8'))

        if not self._is_encrypted(value):
            padding = self._get_padding(value)
            if padding > 0:
                value += bytes("\0".encode('utf-8')) + bytes(
                    ''.encode('utf-8')).join([
                        bytes(random.choice(
                            string.printable).encode('utf-8'))
                        for index in range(padding - 1)])
            if self.block_type != DEFAULT_BLOCK_TYPE:
                self.cipher = self.cipher_object.new(
                    self.secret_key,
                    getattr(self.cipher_object, self.block_type),
                    self.iv)
                value = self.prefix + binascii.b2a_hex(
                    self.iv + self.cipher.encrypt(value)).decode('utf-8')
            else:
                value = self.prefix + binascii.b2a_hex(
                    self.cipher.encrypt(value)).decode('utf-8')
        return value

    def deconstruct(self):
        original = super(BaseEncryptedField, self).deconstruct()
        kwargs = original[-1]
        if self.cipher_type != 'AES':
            kwargs['cipher'] = self.cipher_type
        if self.block_type is not None:
            kwargs['block_type'] = self.block_type
        if self.original_max_length != 40:
            kwargs['max_length'] = self.original_max_length
        return original[:-1] + (kwargs,)


class EncryptedTextField(BaseEncryptedField):

    def get_internal_type(self):
        return 'TextField'

    def formfield(self, **kwargs):
        defaults = {'widget': forms.Textarea}
        defaults.update(kwargs)
        return super(EncryptedTextField, self).formfield(**defaults)


class EncryptedCharField(BaseEncryptedField):

    def get_internal_type(self):
        return "CharField"

    def formfield(self, **kwargs):
        defaults = {'max_length': self.max_length}
        defaults.update(kwargs)
        return super(EncryptedCharField, self).formfield(**defaults)

    def get_db_prep_value(self, value, connection=None, prepared=False):
        if value is not None and not self._is_encrypted(value):
            if len(value) > self.unencrypted_length:
                raise ValueError(
                    "Field value longer than max allowed: " +
                    str(len(value)) + " > " + str(self.unencrypted_length)
                )
        return super(EncryptedCharField, self).get_db_prep_value(
            value,
            connection=connection,
            prepared=prepared,
        )


class BaseEncryptedDateField(BaseEncryptedField):
    # Do NOT define a __metaclass__ for this - it's an abstract parent
    # for EncryptedDateField and EncryptedDateTimeField.
    # If you try to inherit from a class with a __metaclass__, you'll
    # get a very opaque infinite recursion in contribute_to_class.

    def __init__(self, *args, **kwargs):
        kwargs['max_length'] = self.max_raw_length
        super(BaseEncryptedDateField, self).__init__(*args, **kwargs)

    def get_internal_type(self):
        return 'CharField'

    def formfield(self, **kwargs):
        defaults = {'widget': self.form_widget, 'form_class': self.form_field}
        defaults.update(kwargs)
        return super(BaseEncryptedDateField, self).formfield(**defaults)

    def to_python(self, value):
        return self.from_db_value(value)

    def from_db_value(self, value, expression, connection):
        # value is either a date or a string in the format "YYYY:MM:DD"

        if value in EMPTY_VALUES:
            date_value = value
        else:
            if isinstance(value, self.date_class):
                date_value = value
            else:
                date_text = super(BaseEncryptedDateField, self).from_db_value(
                    value, expression, connection)
                date_value = self.date_class(*map(int, date_text.split(':')))
        return date_value

    def get_db_prep_value(self, value, connection=None, prepared=False):
        # value is a date_class.
        # We need to convert it to a string in the format "YYYY:MM:DD"
        if value:
            date_text = value.strftime(self.save_format)
        else:
            date_text = None
        return super(BaseEncryptedDateField, self).get_db_prep_value(
            date_text,
            connection=connection,
            prepared=prepared
        )


class EncryptedDateField(BaseEncryptedDateField):
    form_widget = forms.DateInput
    form_field = forms.DateField
    save_format = "%Y:%m:%d"
    date_class = datetime.date
    max_raw_length = 10  # YYYY:MM:DD


class EncryptedDateTimeField(BaseEncryptedDateField):
    # FIXME:  This doesn't handle time zones, but Python doesn't really either.
    form_widget = forms.DateTimeInput
    form_field = forms.DateTimeField
    save_format = "%Y:%m:%d:%H:%M:%S:%f"
    date_class = datetime.datetime
    max_raw_length = 26  # YYYY:MM:DD:hh:mm:ss:micros


class BaseEncryptedNumberField(BaseEncryptedField):
    # Do NOT define a __metaclass__ for this - it's abstract.
    # See BaseEncryptedDateField for full explanation.
    def __init__(self, *args, **kwargs):
        if self.max_raw_length:
            kwargs['max_length'] = self.max_raw_length
        super(BaseEncryptedNumberField, self).__init__(*args, **kwargs)

    def get_internal_type(self):
        return 'CharField'

    def to_python(self, value):
        return self.from_db_value(value)

    def from_db_value(self, value, expression, connection):
        # value is either an int or a string of an integer
        if isinstance(value, self.number_type) or value == '':
            number = value
        else:
            number_text = super(BaseEncryptedNumberField, self).from_db_value(
                value, expression, connection)
            number = self.number_type(number_text)
        return number

    # def get_prep_value(self, value):
    def get_db_prep_value(self, value, connection=None, prepared=False):
        number_text = self.format_string % value
        return super(BaseEncryptedNumberField, self).get_db_prep_value(
            number_text,
            connection=connection,
            prepared=prepared,
        )


class EncryptedIntField(BaseEncryptedNumberField):
    max_raw_length = len(str(-sys.maxsize - 1))
    number_type = int
    format_string = "%d"


class EncryptedLongField(BaseEncryptedNumberField):
    max_raw_length = None  # no limit
    number_type = int
    format_string = "%d"

    def get_internal_type(self):
        return 'TextField'


class EncryptedFloatField(BaseEncryptedNumberField):
    max_raw_length = 150  # arbitrary, but should be sufficient
    number_type = float
    # If this format is too long for some architectures, change it.
    format_string = "%0.66f"


class PickleField(models.TextField):
    editable = False
    serialize = False

    def get_db_prep_value(self, value, connection=None, prepared=False):
        # We convert data to base64 to prevent errors when
        # unpickling.
        val = codecs.encode(pickle.dumps(value), 'base64').decode()
        return val

    def to_python(self, value):
        return self.from_db_value(value)

    def from_db_value(self, value, expression, connection):
        if not isinstance(value, str):
            return value

        # Tries to convert unicode objects to string, cause loads pickle from
        # unicode excepts ugly ``KeyError: '\x00'``.
        try:
            # Data are in base64 to prevent errors when
            # unpickling.
            val = pickle.loads(codecs.decode(value.encode(), "base64"))
            return val
        # If pickle could not loads from string it's means that it's Python
        # string saved to PickleField.
        except ValueError:
            return value
        except EOFError:
            return value


class EncryptedUSPhoneNumberField(BaseEncryptedField):
    def get_internal_type(self):
        return "CharField"

    def formfield(self, **kwargs):
        try:
            from localflavor.us.forms import USPhoneNumberField
        except ImportError:
            from django.contrib.localflavor.us.forms import USPhoneNumberField

        defaults = {'form_class': USPhoneNumberField}
        defaults.update(kwargs)
        return super(EncryptedUSPhoneNumberField, self).formfield(**defaults)


class EncryptedUSSocialSecurityNumberField(BaseEncryptedField):
    def get_internal_type(self):
        return "CharField"

    def formfield(self, **kwargs):
        try:
            from localflavor.us.forms import USSocialSecurityNumberField
        except ImportError:
            from django.contrib.localflavor.us.forms import USSocialSecurityNumberField

        defaults = {'form_class': USSocialSecurityNumberField}
        defaults.update(kwargs)
        return super(EncryptedUSSocialSecurityNumberField, self).formfield(**defaults)


class EncryptedEmailField(BaseEncryptedField):
    description = _("E-mail address")

    def get_internal_type(self):
        return "CharField"

    def formfield(self, **kwargs):
        from django.forms import EmailField
        defaults = {'form_class': EmailField, 'max_length': self.unencrypted_length}
        defaults.update(kwargs)
        return super(EncryptedEmailField, self).formfield(**defaults)
